README.md
## Fenster
Fenster is a fronted for RDF quad-stores.

It is inspired by, and similar to [Pubby](http://wifo5-03.informatik.uni-mannheim.de/pubby/), but differs in that it shows triples from all public graphs, not just the default graph.

Example of how a resolvable URI, [http://data.deichman.no/resource/tnr_1140686](http://data.deichman.no/resource/tnr_1140686), is presented in Fenster:
![screenshot](https://dl.dropboxusercontent.com/u/27551242/azur.png)

### Status
Fenster is stable and has been in production since November 2013 as a frontend for the [RDF-catalogue](http://data.deichman.no) of [Oslo public library](http://www.deichman.no). Currently it's only been tested against Virtuoso, but presumably any compliant SPARQL endpoint should work. Please let us know if you run into any issues.

### Deployment
Fenster is written in Go, so you'll need the [Go toolchain](http://golang.org/doc/install) in order to build. It compiles to a statically linked binary, so deployment couldn't be simpler:

1. `git clone https://github.com/knakk/fenster`
2. `cd fenster`
3. `make package`*
4. copy `fenster.tar.gz` to your server
5. unpack, adjust settings in `config.ini` and run `fenster`

*You have to do the build step manually if your target platform is of a different architecture than your compilation platform.
[See this guide](http://dave.cheney.net/2012/09/08/an-introduction-to-cross-compilation-with-go) if you don't know where to start.

If you're on Ubuntu, you might want to deploy Fenster as an Upstart service. Example config:
```upstart
description "Fenster"

start on filesystem or runlevel [2345]
stop on runlevel [!2345]

respawn

chdir /path/to/fenster
exec ./fenster
```

#### Apache routing
If Fenster is running on same server as the RDF-store, you'll have to proxy the requests to the SPARQL endpoint.

Here is an example Apache config, given Fenster running on localhost:8080 and SPARQL endpoint running on localhost:8890/sparql:

```apache
<VirtualHost *:80>

    ServerAdmin serveradmin@example.no
    DocumentRoot /var/www/example.com
    ServerName example.com

    ProxyRequests off
    ProxyPreserveHost on
    ProxyTimeout        300
    # Proxy ACL
    <Proxy *>
        Order allow,deny
        Allow from all
    </Proxy>
    RewriteEngine on
    RewriteRule ^/sparql(.*)$ http://localhost:8890/sparql$1 [P]

    # default proxy if not handled above
    ProxyPass / http://example.com:8080/ timeout=300
    ProxyPassReverse / http://example.com:8080/

</VirtualHost>
```


### License
GPLv3


CHANGELOG
0.3   26.07.2014
==================================================
* Support gzip-compression.
* Optionally fetch resource literals on mouse hover.
* Expose status and metrics in ./status endpoint.
* Display total count of solutions on resource,
  if larger than the number of solutions fetched.

0.2   07.10.2013
===================================================
* First public release

utils_test.go
package main

import "testing"

func TestPrefixify(t *testing.T) {
	prefixes := [][]string{
		{"dc", "http://purl.org/dc/terms/"},
		{"foaf", "http://xmlns.com/foaf/0.1/"}}

	var (
		u1 = "<http://purl.org/dc/terms/title>"
		u2 = "http://xmlns.com/foaf/0.1/name"
		u3 = "<http://purl.org/dc/terms/deeper/path>"
	)

	var resTests = []struct {
		in  interface{}
		out interface{}
	}{
		{prefixify(&prefixes, u1), "dc:title"},
		{prefixify(&prefixes, u2), "foaf:name"},
		{prefixify(&prefixes, u3), u3},
	}

	for i, tt := range resTests {
		if tt.in != tt.out {
			t.Errorf("%d) expected %v, got %v", i, tt.out, tt.in)
		}
	}
}


browser_test.js
casper.test.begin('Existing resource', 9, function suite(test) {
  casper.start('http://localhost:8080/resource/tnr_1140686', function() {
    test.assertHttpStatus(200, 'response status code 200');
    test.assertTitle('Azur', 'title as expected');
  });

  casper.thenOpen('http://localhost:8080/resource/tnr_1140686.html', function(response) {
    test.assertHttpStatus(200, 'response status code 200');
    test.assertMatch(response.headers.get('Content-Type'), /^text\/html/, 'correct content-type');
    test.assertTitle('Azur', 'title as expected');
  });

  casper.thenOpen('http://localhost:8080/resource/tnr_1140686.json', function(response) {
    test.assertMatch(response.headers.get('Content-Type'), /^application\/json/, 'correct content-type');
  });

  casper.thenOpen('http://localhost:8080/resource/tnr_1140686.rdf', function(response) {
    test.assertMatch(response.headers.get('Content-Type'), /^application\/x-trig/, 'correct content-type');
  });

  casper.thenOpen('http://localhost:8080/resource/tnr_1140686.zappa', function() {
    test.assertHttpStatus(400, 'status 400 on bad request');
    this.test.assertTextExists('Unsupported output format', 'error message in page body');
  });

  casper.run(function() {
    test.done();
  });
});

casper.test.begin('Missing resource', 1, function suite(test) {
  casper.start('http://localhost:8080/an/very/unlikely/path/doh', function() {
    test.assertHttpStatus(404, 'response status code 404');
  }).run(function() {
    test.done();
  });
});

fenster_test.go
package main

import (
	"io/ioutil"
	"net/http"
	"net/http/httptest"
	"regexp"
	"testing"
)

func TestAppnameAndVersionInPage(t *testing.T) {
	handler := &mainHandler{}
	server := httptest.NewServer(handler)
	defer server.Close()

	resp, err := http.Get(server.URL + "/resource/tnr_1140686")
	if err != nil {
		t.Fatal(err)
	}

	if resp.StatusCode != 200 {
		t.Fatalf("Got non-200 response: %d\n", resp.StatusCode)
	}

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		t.Fatal(err)
	}

	if !regexp.MustCompile(`Fenster version (\d\.)+`).MatchString(string(body)) {
		t.Error("Application name and version missing.")
	}
}


tee.go
package main

// This file is stolen from:
// https://github.com/rcrowley/go-tigertonic/blob/master/tee.go

import (
	"bytes"
	"net/http"
)

// TeeHeaderResponseWriter is an http.ResponseWriter that both writes and
// records the response status and headers for post-processing.
type TeeHeaderResponseWriter struct {
	http.Flusher
	http.ResponseWriter
	StatusCode int
}

// NewTeeHeaderResponseWriter constructs a new TeeHeaderResponseWriter that
// writes responses through another http.ResponseWriter and records the
// response status and headers for post-processing.
func NewTeeHeaderResponseWriter(w http.ResponseWriter) *TeeHeaderResponseWriter {
	return &TeeHeaderResponseWriter{ResponseWriter: w}
}

// Flush implements the http.Flusher interface, if possible, to support streaming
// responses to clients.
func (w *TeeHeaderResponseWriter) Flush() {
	if f, ok := w.ResponseWriter.(http.Flusher); ok {
		f.Flush()
	}
}

// WriteHeader writes the response line and headers to the client via the
// underlying http.ResponseWriter and records the status for post-processing.
func (w *TeeHeaderResponseWriter) WriteHeader(code int) {
	w.ResponseWriter.WriteHeader(code)
	w.StatusCode = code
}

// TeeResponseWriter is an http.ResponseWriter that both writes and records the
// response status, headers, and body for post-processing.
type TeeResponseWriter struct {
	http.Flusher
	http.ResponseWriter
	Body       bytes.Buffer
	StatusCode int
}

// NewTeeResponseWriter constructs a new TeeResponseWriter that writes
// responses through another http.ResponseWriter and records the response
// status, headers, and body for post-processing.
func NewTeeResponseWriter(w http.ResponseWriter) *TeeResponseWriter {
	return &TeeResponseWriter{ResponseWriter: w}
}

// Flush implements the http.Flusher interface, if possible, to support streaming
// responses to clients.
func (w *TeeResponseWriter) Flush() {
	if f, ok := w.ResponseWriter.(http.Flusher); ok {
		f.Flush()
	}
}

// Write writes the byte slice to the client via the underlying
// http.ResponseWriter and records it for post-processing.
func (w *TeeResponseWriter) Write(p []byte) (int, error) {
	if n, err := w.ResponseWriter.Write(p); nil != err {
		return n, err
	}
	return w.Body.Write(p)
}

// WriteHeader writes the response line and headers to the client via the
// underlying http.ResponseWriter and records the status for post-processing.
func (w *TeeResponseWriter) WriteHeader(code int) {
	w.ResponseWriter.WriteHeader(code)
	w.StatusCode = code
}


utils.go
package main

import (
	"errors"
	"fmt"
	"html/template"
	"io"
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"time"

	"github.com/knakk/rdf"
	"github.com/mreiferson/go-httpclient"
)

type remoteRepo struct {
	endpoint string
	client   *http.Client
}

func newRepo(endpoint string, openTimeout, readTimeout time.Duration) *remoteRepo {
	transport := &httpclient.Transport{
		ConnectTimeout:        openTimeout,
		RequestTimeout:        openTimeout + readTimeout,
		ResponseHeaderTimeout: readTimeout,
	}
	client := &http.Client{Transport: transport}
	return &remoteRepo{endpoint: endpoint, client: client}
}

func (r *remoteRepo) Close() {
	//r.client.Transport.Close()
}

// Query sends a request to a remote SPARQL endpoint and returns the unparsed
// response body
func (r *remoteRepo) Query(endpoint string, query string, format string) (io.ReadCloser, error) {
	reqDefaults := url.Values{}
	reqDefaults.Set("query", query)

	switch format {
	case "json":
		reqDefaults.Set("format", "application/sparql-results+json")
	case "rdf":
		reqDefaults.Set("format", "application/x-trig")
	default:
		reqDefaults.Set("format", "application/sparql-results+json")
	}

	req, err := http.NewRequest("POST", fmt.Sprintf("%s?%v", endpoint, reqDefaults.Encode()), nil)
	if err != nil {
		return nil, fmt.Errorf("error preparing http request: %v", err)
	}

	resp, err := r.client.Do(req)
	if err != nil {
		// trim URL from error message
		i := strings.Index(err.Error(), "dial")
		if i != -1 {
			return nil, errors.New(err.Error()[i:])
		}
		return nil, err
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("SPARQL endpoint responded with HTTP status code: %v", resp.StatusCode)
	}

	return resp.Body, nil
}

// prefixify returns the prefixed form of an URI if the prefix and namespaces
// is found in the prefixes array, which must have the following form:
// ["dc", "http://purl.org/dc/terms/"], ["foaf", "http://xmlns....etc"]
//
//It returns the string unmodified if no match is found.
func prefixify(prefixes *[][]string, uri string) string {
	uriOriginal := uri

	// trim enclosing < > if present
	if strings.HasPrefix(uri, "<") && strings.HasSuffix(uri, ">") {
		uri = strings.TrimSuffix(uri, ">")[1:]
	}

	for _, prefixPair := range *prefixes {
		if strings.HasPrefix(uri, prefixPair[1]) {
			//avoid prefixing URI's witch has forward slash after the namespaced part
			if regexp.MustCompile(`\/`).MatchString(strings.TrimPrefix(uri, prefixPair[1])) {
				return uriOriginal
			}
			return strings.Replace(uri, prefixPair[1], prefixPair[0]+":", 1)
		}
	}
	return uriOriginal
}

func rejectWhereEmpty(key string, solutions []map[string]rdf.Term) []map[string]interface{} {
	// TODO clean up this function; choose another name too..
	included := make([]map[string]interface{}, 1)
	for _, m := range solutions {
		if m[key] != nil {
			tm := make(map[string]interface{})
			for k, v := range m {
				if k != "g" && k != "p" && strings.HasPrefix(v.String(), "<"+conf.BaseURI) {
					var link string
					if conf.UI.FetchLiterals {
						link = fmt.Sprintf("<div class='relative'><a class=\"resource-link\" href='/%v'>%v</a><div class=\"tooltip\"><strong>%s</strong><div class='literals'>...</div></div></div>",
							v.String()[25:len(v.String())-1], template.HTMLEscapeString(v.String()), template.HTMLEscapeString(v.String()))
					} else {
						link = fmt.Sprintf("<a href='/%v'>%v</a>",
							v.String()[25:len(v.String())-1], template.HTMLEscapeString(v.String()))
					}

					// TODO magic number 25 = len(conf.BaseURI)??
					tm[k] = template.HTML(link)
				} else {
					if conf.Vocab.Enabled {
						tm[k] = prefixify(&conf.Vocab.Dict, v.String())
					} else {
						tm[k] = v.String()
					}
				}
			}
			included = append(included, tm)
		}
	}
	return included
}

// findTitle iterates over solutions and returns the first literal string where the RDF
// predicate matches any of the predicates in titlePredicates, or an empty string
// if none is found.
func findTitle(titlePredicates []string, solutions []map[string]rdf.Term) string {
	if len(titlePredicates) == 0 {
		return ""
	}

	for _, m := range solutions {
		for _, p := range titlePredicates {
			if m["p"].String() == "<"+p+">" {
				return m["o"].String()
			}
		}
	}
	return ""
}

// findImages iterates over solutions and returns any images, that is,
// objects where the predicate is one of the predicates slice.
func findImages(predicates []string, solutions []map[string]rdf.Term) []string {
	var images []string
	if !conf.UI.ShowImages {
		return images
	}
	for _, m := range solutions {
		for _, p := range predicates {
			if m["p"].String() == "<"+p+">" {
				images = append(images, strings.TrimSuffix(m["o"].String()[1:], ">"))
				if len(images) == conf.UI.NumImages {
					return images
				}
			}
		}
	}
	return images
}


metrics.go
package main

import (
	"fmt"
	"net/http"
	"os"
	"time"

	"github.com/rcrowley/go-metrics"
)

type appMetrics struct {
	StartTime time.Time
	PID       int
}

type exportMetrics struct {
	UpTime  string
	PID     int
	Metrics metrics.Registry
}

func registerMetrics() *appMetrics {
	var m appMetrics

	m.StartTime = time.Now()
	m.PID = os.Getpid()

	return &m
}

func (m *appMetrics) Export() *exportMetrics {
	now := time.Now()
	uptime := now.Sub(m.StartTime)

	return &exportMetrics{
		UpTime:  uptime.String(),
		PID:     m.PID,
		Metrics: metrics.DefaultRegistry,
	}
}

// The following code is from copied from
// https://github.com/rcrowley/go-tigertonic/blob/master/metrics.go

// CounterByStatusXX is an http.Handler that counts responses by the first
// digit of their HTTP status code via go-metrics.
type CounterByStatusXX struct {
	counter1xx, counter2xx, counter3xx, counter4xx, counter5xx metrics.Counter
	handler                                                    http.Handler
}

// CountedByStatusXX returns an http.Handler that passes requests to an
// underlying http.Handler and then counts the response by the first digit of
// its HTTP status code via go-metrics.
func CountedByStatusXX(
	handler http.Handler,
	name string,
	registry metrics.Registry,
) *CounterByStatusXX {
	if nil == registry {
		registry = metrics.DefaultRegistry
	}
	c := &CounterByStatusXX{
		counter1xx: metrics.NewCounter(),
		counter2xx: metrics.NewCounter(),
		counter3xx: metrics.NewCounter(),
		counter4xx: metrics.NewCounter(),
		counter5xx: metrics.NewCounter(),
		handler:    handler,
	}
	if err := registry.Register(
		fmt.Sprintf("%s-1xx", name),
		c.counter1xx,
	); nil != err {
		panic(err)
	}
	if err := registry.Register(
		fmt.Sprintf("%s-2xx", name),
		c.counter2xx,
	); nil != err {
		panic(err)
	}
	if err := registry.Register(
		fmt.Sprintf("%s-3xx", name),
		c.counter3xx,
	); nil != err {
		panic(err)
	}
	if err := registry.Register(
		fmt.Sprintf("%s-4xx", name),
		c.counter4xx,
	); nil != err {
		panic(err)
	}
	if err := registry.Register(
		fmt.Sprintf("%s-5xx", name),
		c.counter5xx,
	); nil != err {
		panic(err)
	}
	return c
}

// ServeHTTP passes the request to the underlying http.Handler and then counts
// the response by its HTTP status code via go-metrics.
func (c *CounterByStatusXX) ServeHTTP(w0 http.ResponseWriter, r *http.Request) {
	w := NewTeeHeaderResponseWriter(w0)
	c.handler.ServeHTTP(w, r)
	if w.StatusCode < 200 {
		c.counter1xx.Inc(1)
	} else if w.StatusCode < 300 {
		c.counter2xx.Inc(1)
	} else if w.StatusCode < 400 {
		c.counter3xx.Inc(1)
	} else if w.StatusCode < 500 {
		c.counter4xx.Inc(1)
	} else {
		c.counter5xx.Inc(1)
	}
}

// Timer is an http.Handler that counts requests via go-metrics.
type Timer struct {
	metrics.Timer
	handler http.Handler
}

// Timed returns an http.Handler that starts a timer, passes requests to an
// underlying http.Handler, stops the timer, and updates the timer via
// go-metrics.
func Timed(handler http.Handler, name string, registry metrics.Registry) *Timer {
	timer := &Timer{
		Timer:   metrics.NewTimer(),
		handler: handler,
	}
	if nil == registry {
		registry = metrics.DefaultRegistry
	}
	if err := registry.Register(name, timer); nil != err {
		panic(err)
	}
	return timer
}

// ServeHTTP starts a timer, passes the request to the underlying http.Handler,
// stops the timer, and updates the timer via go-metrics.
func (t *Timer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	defer t.UpdateSince(time.Now())
	t.handler.ServeHTTP(w, r)
}


rdf/rdf.go
package rdf

import (
	"fmt"
	"regexp"
	"strings"
	"time"
)

type TermType int

// RDF Term types
const (
	BLANK TermType = iota
	LITERAL
	URI
)

// Term is the common interface for RDF Terms
type Term interface {
	String() string
	Type() TermType
	Eq(Term) bool
	Value() interface{}
}

// Uri RDF Term
type Uri string

// String returns the string representation of an Uri
func (u Uri) String() string {
	return "<" + string(u) + ">"
}

// Type returns the RDF Term type (URI)
func (u Uri) Type() TermType {
	return URI
}

// Value returns the Uri as value, i.e the Uri string without the enclosing < >
func (u Uri) Value() interface{} {
	return string(u)
}

// Eq tests the equality of an Uri against another RDF Term
func (u Uri) Eq(t Term) bool {
	if u.Type() == t.Type() {
		if u.String() == t.String() {
			return true
		}
	}
	return false
}

// NewUri returns a new Uri.
// Equivalent of var x Uri = "{{uri}}"
func NewUri(s string) Uri {
	return Uri(s)
}

// Literal RDF Type
type Literal struct {
	Lang     string
	Val      interface{}
	DataType Uri
}

// String returns the string representation of a Literal
func (l Literal) String() string {
	if l.Lang != "" {
		return fmt.Sprintf("%v@%s", l.Val, l.Lang)
	}
	if l.DataType != "" {
		return fmt.Sprintf("%v^^%v", l.Val, l.DataType)
	}
	return fmt.Sprintf("%v", l.Val)
}

// Type returns the RDF Term type (LITERAL)
func (l Literal) Type() TermType {
	return LITERAL
}

// Value returns the Literal value, in the corresponding go type, i.e.
// xsd:integer -> int, xsd:float -> float64 and so on.
//
// For language-tagged literals, it returns the string without the trailing
// language tag.
func (l Literal) Value() interface{} {
	return l.Val
}

// Eq tests the equality of a Literal against another RDF Term
func (l Literal) Eq(t Term) bool {
	if l.Type() == t.Type() {
		if l.String() == t.String() {
			return true
		}
	}
	return false
}

// NewLiteral is a Literal constructor.
// It sets the DataType based on it's input, or returns an error if it fails to
// guess the DataType.
// Suffix with @ to create a language tagged literal, ex: NewLiteral("string@en")
// If Lang is provided, DataType is ignored.
// If you need a custom DataType, you must create the literal with the normal
// struct syntax, ex:
// l := Literal{Val: "my-val", DataType: Uri("my-custom-type")}
func NewLiteral(v interface{}) (Literal, error) {
	switch t := v.(type) {
	default:
		return Literal{}, fmt.Errorf("cannot infer xsd:datatype from %v", t)
	case bool:
		return Literal{Val: t, DataType: Uri("http://www.w3.org/2001/XMLSchema#boolean")}, nil
	case int:
		return Literal{Val: t, DataType: Uri("http://www.w3.org/2001/XMLSchema#int")}, nil
	case string:
		r := regexp.MustCompile(`@[a-z]{2,3}$`)
		if r.MatchString(t) {
			// matches a iso639-2 language code:
			// http://www.loc.gov/standards/iso639-2/php/code_list.php
			tag := r.FindString(t)
			return Literal{Val: strings.Split(t, tag)[0], Lang: tag[1:]}, nil
		}
		return Literal{Val: t, DataType: Uri("http://www.w3.org/2001/XMLSchema#string")}, nil
	case float64:
		return Literal{Val: t, DataType: Uri("http://www.w3.org/2001/XMLSchema#float")}, nil
	case time.Time:
		return Literal{}, fmt.Errorf("not implemented")
		/*    ti, err := time.Parse(time.RFC3339, t)
		      if err == nil {
		        // matches a RFC3339 date, assuming xsd:dateTime
		        return Literal{Val: ti, DataType: Uri("http://www.w3.org/2001/XMLSchema#dateTime")}, nil
		      }*/
	}
}

// Blank Node RDF Term.
//
// Note: It is left to the consumer to enforce a globaly unique identifier.
type Blank string

// String returns the string representation of a blank node
func (b Blank) String() string {
	return "_:" + string(b)
}

// Type returns the RDF Term type (BLANK)
func (b Blank) Type() TermType {
	return BLANK
}

// Value returns the value of a Blank Node (which is nil)
func (b Blank) Value() interface{} {
	return nil
	// or empty string?
}

// Eq tests the equality of a Blank node against another RDF Term
func (b Blank) Eq(t Term) bool {
	if b.Type() == t.Type() {
		if b.String() == t.String() {
			return true
		}
	}
	return false
}

// NewBlank returns a new Blank Node.
func NewBlank(s string) Blank {
	if s == "" {
		return Blank(fmt.Sprint(&s))
	}
	return Blank(s)
}

// Triple represents a RDF statement.
type Triple struct {
	Subject, Predicate, Object Term
}

// Quad represents a RDF statement within a specified graph.
type Quad struct {
	Graph                      Uri
	Subject, Predicate, Object Term
}

// TODO serialization?
// At least Turtle & N3

/*
type TripleGraph []Triple

type QuadGraph []Quad

*/


fenster.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"html/template"
	"io"
	"log"
	"net/http"
	"regexp"
	"strings"
	"time"

	"github.com/BurntSushi/toml"
	"github.com/gorilla/handlers"
	"github.com/knakk/rdf"
	"github.com/knakk/sparql"
	"github.com/oxtoacart/bpool"
	"github.com/rcrowley/go-metrics"
)

const (
	version = "0.3"
	queries = `
# tag: select
SELECT *
WHERE { GRAPH ?g { { <{{.URI}}> ?p ?o } UNION { ?s ?p <{{.URI}}> } } }
LIMIT {{.Limit}}

# tag: count
SELECT COUNT(?s) AS ?maxO, COUNT(?o) as ?maxS
WHERE { GRAPH ?g { { <{{.URI}}> ?p ?o } UNION { ?s ?p <{{.URI}}> } } }

# tag: construct
CONSTRUCT { GRAPH ?g { <{{.URI}}> ?p ?o . ?s ?p <{{.URI}}> } }
WHERE { GRAPH ?g { { <{{.URI}}> ?p ?o } UNION { ?s ?p <{{.URI}}> } } }

#tag: literals
SELECT DISTINCT ?p, ?o
WHERE { <{{.URI}}> ?p ?o .
        FILTER isLiteral(?o) }
`
)

var (
	templates = template.Must(template.ParseFiles("data/html/index.html", "data/html/error.html"))
	conf      Config
	repo      *remoteRepo
	status    *appMetrics
	qBank     sparql.Bank
	suffixRg  = regexp.MustCompile(`\.[a-z1-9]+$`)
	bufpool   *bpool.BufferPool
)

type mainHandler struct{}

// rdfHandler serves the quads in TriG syntax
// http://wifo5-03.informatik.uni-mannheim.de/bizer/trig/
func rdfHandler(w http.ResponseWriter, r *http.Request) {
	uri := conf.BaseURI + strings.TrimSuffix(r.URL.Path, ".rdf")
	format := "rdf"

	q, _ := qBank.Prepare("construct", struct{ URI string }{uri})
	resp, err := repo.Query(conf.QuadStore.Endpoint, q, format)
	if err != nil {
		errorHandler(w, r, err.Error()+". Refresh to try again.\n\nYou can increase the timeout values in Fensters configuration file.", http.StatusInternalServerError)
		return
	}
	defer resp.Close()

	w.Header().Set("Content-Type", "application/x-trig")
	io.Copy(w, resp)
}

// jsonHandler serves the raw "application/sparql-results+json" results from
// the SPARQL endpoint
func jsonHandler(w http.ResponseWriter, r *http.Request) {
	uri := conf.BaseURI + strings.TrimSuffix(r.URL.Path, ".json")
	format := "json"

	q, _ := qBank.Prepare("select",
		struct {
			URI   string
			Limit int
		}{uri, conf.QuadStore.ResultsLimit})

	resp, err := repo.Query(conf.QuadStore.Endpoint, q, format)
	if err != nil {
		errorHandler(w, r,
			err.Error()+`. Refresh to try again.\n\n
			You can increase the timeout values in Fensters configuration file.`,
			http.StatusInternalServerError)
		return
	}
	defer resp.Close()

	w.Header().Set("Content-Type", "application/json")
	io.Copy(w, resp)
}

// mainHandler serves the resource HTML presentation, or dispatches to the
// rdfHandler or jsonHandler
func (m mainHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	if r.URL.Path == "/" {
		// redirect from root to info page
		http.Redirect(w, r, conf.UI.RootRedirectTo, http.StatusFound)
		return
	}

	var uri string
	resolved := false
	suffix := suffixRg.FindString(r.URL.Path)

	switch suffix {
	case "":
		break
	case ".html":
		uri = conf.BaseURI + strings.TrimSuffix(r.URL.Path, ".html")
		resolved = true
	case ".json":
		jsonHandler(w, r)
		return
	case ".rdf":
		rdfHandler(w, r)
		return
	default:
		errorHandler(w, r,
			fmt.Sprintf("Unsupported output format: %s.\n\n"+
				"Valid formats are: html, json, rdf", suffix[1:]), http.StatusBadRequest)
		return
	}

	// The URI should be an exclusive identifier of the resource; so we redirect
	// to URI+.html
	if !resolved {
		http.Redirect(w, r, r.URL.Path+".html", http.StatusFound)
		return
	}

	q, _ := qBank.Prepare("select",
		struct {
			URI   string
			Limit int
		}{uri, conf.QuadStore.ResultsLimit})

	resp, err := repo.Query(conf.QuadStore.Endpoint, q, "json")
	if err != nil {
		//println(err.Error())
		errorHandler(w, r,
			err.Error()+". Refresh to try again.\n\nYou can increase the timeout"+
				" values in Fensters configuration file.", http.StatusInternalServerError)
		return
	}
	defer resp.Close()

	res, err := sparql.ParseJSON(resp)
	if err != nil {
		errorHandler(w, r,
			"Failed to parse JSON response from remote SPARQL endpoint.",
			http.StatusInternalServerError)
		return
	}

	if len(res.Results.Bindings) == 0 {
		errorHandler(w, r, "This URI has no information", http.StatusNotFound)
		return
	}

	var maxS, maxO int
	if len(res.Results.Bindings) >= conf.QuadStore.ResultsLimit {
		// Fetch solution counts, if we hit the results limit
		q, _ := qBank.Prepare("count", struct{ URI string }{uri})
		resp, err := repo.Query(conf.QuadStore.Endpoint, q, "json")
		if err == nil {
			res, err := sparql.ParseJSON(resp)
			if err == nil {
				b := res.Bindings()
				maxS = b["maxS"][0].(*rdf.Literal).Value.(int)
				maxO = b["maxO"][0].(*rdf.Literal).Value.(int)
			}
			resp.Close()
		}

	}

	solutions := res.Solutions()
	subj := rejectWhereEmpty("o", solutions)
	obj := rejectWhereEmpty("s", solutions)
	data := struct {
		Title               string
		License, LicenseURL string
		Endpoint            string
		Name, Version, URI  string
		AsSubject           []map[string]interface{}
		AsObject            []map[string]interface{}
		AsSubjectSize       int
		AsObjectSize        int
		MaxSubject          int
		MaxObject           int
		Images              []string
	}{
		findTitle(conf.UI.TitlePredicates, solutions),
		conf.License,
		conf.LicenseURL,
		conf.QuadStore.Endpoint,
		"Fenster",
		string(version),
		uri,
		subj,
		obj,
		len(subj) - 1,
		len(obj) - 1,
		maxS,
		maxO,
		findImages(conf.UI.ImagePredicates, solutions),
	}

	buf := bufpool.Get()
	defer bufpool.Put(buf)
	err = templates.ExecuteTemplate(buf, "index.html", data)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	w.WriteHeader(http.StatusOK)
	buf.WriteTo(w)
}

// errorHandler serves 40x & 50x error pages
func errorHandler(w http.ResponseWriter, r *http.Request, msg string, status int) {
	data := struct {
		ErrorCode int
		ErrorMsg  string
	}{
		status,
		msg,
	}

	buf := bufpool.Get()
	defer bufpool.Put(buf)
	err := templates.ExecuteTemplate(buf, "error.html", data)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	w.WriteHeader(status)
	buf.WriteTo(w)
}

// serveFile serves a single file from disk
func serveFile(filename string) func(w http.ResponseWriter, r *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) {
		w.Header().Add("Cache-Control", "max-age=2629743, public")
		http.ServeFile(w, r, filename)
	}
}

func statusHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	err := json.NewEncoder(w).Encode(status.Export())
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
	}
}

func literalsHandler(w http.ResponseWriter, r *http.Request) {
	uri := r.FormValue("uri")
	q, _ := qBank.Prepare("literals", struct{ URI string }{uri})
	resp, err := repo.Query(conf.QuadStore.Endpoint, q, "json")
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	res, err := sparql.ParseJSON(resp)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	defer resp.Close()
	if len(res.Results.Bindings) == 0 {
		w.Write([]byte("No literals on resource"))
		return
	}

	var b bytes.Buffer
	b.WriteString("<table class='preview'>")
	for _, s := range res.Solutions() {
		b.WriteString("<tr><td>" + prefixify(&conf.Vocab.Dict, s["p"].String()) + "</td><td>")
		b.WriteString(s["o"].String() + "</td></tr>")
	}
	b.WriteString("</table>")
	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	w.Write(b.Bytes())
}

func main() {
	// Init bufferpool, used for rendering templates
	bufpool = bpool.NewBufferPool(48)

	// Load config file
	if _, err := toml.DecodeFile("config.ini", &conf); err != nil {
		log.Fatal("Couldn't parse config file: ", err)
	}

	// Setup remote repository
	repo = newRepo(
		conf.QuadStore.Endpoint,
		time.Duration(conf.QuadStore.OpenTimeout)*time.Millisecond,
		time.Duration(conf.QuadStore.ReadTimeout)*time.Millisecond,
	)

	// Parse Query bank
	qBank = sparql.LoadBank(bytes.NewBufferString(queries))

	// Register metrics
	status = registerMetrics()

	// HTTP routing
	mux := http.NewServeMux()
	var handler mainHandler
	mux.HandleFunc("/robots.txt", serveFile("data/robots.txt"))
	mux.HandleFunc("/css/styles.css", serveFile("data/css/styles.css"))
	mux.HandleFunc("/favicon.ico", serveFile("data/favicon.ico"))
	mux.HandleFunc("/.status", statusHandler)
	mux.HandleFunc("/literals", literalsHandler)
	mux.Handle("/", Timed(CountedByStatusXX(handler, "status", metrics.DefaultRegistry),
		"responseTime",
		metrics.DefaultRegistry))

	fmt.Printf("Listening on port %d ...\n", conf.ServePort)
	http.ListenAndServe(fmt.Sprintf(":%d", conf.ServePort), handlers.CompressHandler(mux))
}
